Update Scala, dependencies and sbt plugins

guizmaii opened the pull request:
We can't update to scala-native v0.5.x because https://github.com/zio/izumi-reflect/pull/459 is blocked by scalatest

guizmaii committed:
Update Scala, dependencies and sbt plugins

guizmaii committed:
Remove warnings

guizmaii committed:
Rollback scala-native to v.0.4.x

Because of https://github.com/scalatest/scalatest/pull/2318

guizmaii committed:
Fix compilation errors

guizmaii committed:
Fix compilation errors

guizmaii committed:
scalacheck 1.18.0 is only build for scala-native v.0.5.x

guizmaii committed:
Clean ci.yml

guizmaii committed:
Improve `testJvms` CI step to test on 11, 17 and 21

guizmaii committed:
Remove compilation errors

guizmaii committed:
Remove compilation errors

guizmaii committed:
Simplify scala versions declarations

kyri-petrou commented on .github/workflows/ci.yml:
```diff
@@ -152,7 +151,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        scala: ['2.12.18', '2.13.12', '3.3.1']
```
What about Scala 3 to `3.3.3`?

kyri-petrou replied:
Ah nevermind me, just saw that it was changed in the next commit

guizmaii committed:
Unify CI declarations

guizmaii committed:
Fix import

guizmaii committed:
Remove compilation errors

guizmaii committed:
Remove custom sbt script

guizmaii committed:
Update actions/setup-java

guizmaii committed:
Update actions/checkout

guizmaii committed:
Fix CI

guizmaii commented on .github/workflows/ci.yml:
```diff
@@ -2,7 +2,6 @@ name: CI
 
 env:
   JDK_JAVA_OPTIONS: -XX:+PrintCommandLineFlags -Xmx6G -Xss4M -XX:+UseG1GC # JDK_JAVA_OPTIONS is _the_ env. variable to use for modern Java
-  JVM_OPTS: -XX:+PrintCommandLineFlags -Xmx6G -Xss4M -XX:+UseG1GC # for Java 8 only (sadly, it is not modern enough for JDK_JAVA_OPTIONS)
```
We don't support Java8 anymore

guizmaii committed:
Remove compilation errors

guizmaii committed:
Try fix CI error:

```
[warn] In the last 10 seconds, 5.18 (52.6%) were spent in GC. [Heap: 0.00GB free of 1.00GB, max 1.00GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[warn] In the last 10 seconds, 6.369 (67.4%) were spent in GC. [Heap: 0.00GB free of 1.00GB, max 1.00GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.

              total        used        free      shared     buffers       cache   available
Mem:          16768        2589       11296          35         230        2652       13780
Swap:          7516           0        7516
Total:        24284        2589       18812
[error] ## Exception when compiling 16 sources to /home/runner/work/zio/zio/scalafix/input/target/scala-2.13/classes
```

guizmaii committed:
scalafmt

guizmaii committed:
Fix CI

guizmaii committed:
Fix CI

guizmaii committed:
`Compile / fork := false` is declared twice in the `doc` module

guizmaii committed:
Debug sbt command

guizmaii committed:
Put back `JVM_OPTS` envvar

guizmaii committed:
Rollback `Compile / fork` change

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
debug

guizmaii committed:
Use .jvmopts-ci

guizmaii committed:
Prefer `SBT_OPTS`

guizmaii committed:
Debug $SBT_OPTS

guizmaii committed:
Rollback: Debug $SBT_OPTS

guizmaii committed:
clean

guizmaii committed:
Add more debug for future maintainers

guizmaii commented:
@ghostdogpr FYI, we still have a flaky CI because of:
```scala
[error] /home/runner/work/zio/zio/core/shared/src/main/scala/zio/ZLayer.scala:654:50: type arguments [R1,E1,B] do not conform to method flatMap's type parameter bounds [RIn1 <: $this.self.R,E1 >: $this.self.E,ROut2]
[error]           fromLayer[R1, E1, B](self.layer.flatMap[R1, E1, B](a => ZLayer[R1, E1, B](k(a.get[A])))) //
[error]           
```

Any idea how to fix that? 🤔

See https://github.com/zio/zio/actions/runs/8733264034/job/23961640036

ghostdogpr commented:
> @ghostdogpr FYI, we still have a flaky CI because of:
> 
> ```scala
> 
> [error] /home/runner/work/zio/zio/core/shared/src/main/scala/zio/ZLayer.scala:654:50: type arguments [R1,E1,B] do not conform to method flatMap's type parameter bounds [RIn1 <: $this.self.R,E1 >: $this.self.E,ROut2]
> 
> [error]           fromLayer[R1, E1, B](self.layer.flatMap[R1, E1, B](a => ZLayer[R1, E1, B](k(a.get[A])))) //
> 
> [error]           
> 
> ```
> 
> 
> 
> Any idea how to fix that? 🤔

No, I tried adding explicit parameters and I thought it worked because I didn't see it for a while but apparently it's not enough.

guizmaii committed:
clean

guizmaii commented on core-tests/shared/src/test/scala/zio/ExecutorSpec.scala:
```diff
@@ -21,9 +21,9 @@ final class CheckPrintThrowable extends Throwable {
-  val failing = new TestExecutor(false)
-  val y       = new TestExecutor(true)
-  val u       = new TestExecutor(true)
+  val failing = new TestExecutor(submitResult = false)
```
Scala 2.13.13 forces you to use by-name params when using booleans

guizmaii commented on project/BuildHelper.scala:
```diff
@@ -1,29 +1,15 @@
+import sbtcrossproject.CrossPlugin.autoImport.*
 
 object BuildHelper {
-  private val versions: Map[String, String] = {
-
-    import java.util.{List => JList, Map => JMap}
-    val doc = new Load(LoadSettings.builder().build())
-      .loadFromReader(scala.io.Source.fromFile(".github/workflows/ci.yml").bufferedReader())
-    val yaml = doc.asInstanceOf[JMap[String, JMap[String, JMap[String, JMap[String, JMap[String, JList[String]]]]]]]
-    val list = yaml.get("jobs").get("test").get("strategy").get("matrix").get("scala").asScala
-    list.map { v =>
-      val vs = v.split('.'); val init = vs.take(vs(0) match { case "2" => 2; case _ => 1 }); (init.mkString("."), v)
-    }.toMap
-  }
-  val Scala212: String = versions("2.12")
-  val Scala213: String = versions("2.13")
-  val Scala3: String   = versions("3")
```
Needlessly complex since sbt now handles `2.12.x`, etc. versions

guizmaii commented on scalafix/output/src/main/scala/fix/ZIOSpec.scala:
```diff
@@ -11,7 +11,7 @@ import zio.test._
 import scala.annotation.tailrec
 import scala.util.{Failure, Success, Try}
-import zio.{ Clock, FiberId, Random, Random, _ }
+import zio.{ Clock, FiberId, Random, _ }
```
Was imported twice

kyri-petrou replied:
Should this just be `import zio._` or `zio.{ Clock, FiberId, Random }`?

guizmaii replied:
No idea. I just remove an compilation issue when I'm compiling with Intellij

guizmaii commented on sbt:
```diff
@@ -1,664 +0,0 @@
-#!/usr/bin/env bash
```
Not sure why we needed a specific sbt script. Anyone knows?
I don't think that's a great idea. Let's keep the default one if possible.
The CI passes without it

kyri-petrou replied:
I think it's just convenient for people that don't have `sbt` installed so that they can do `./sbt compile`. But I'm all up for removing it, I'm struggling to imagine people that would want to compile this project not having sbt installed

ghostdogpr commented on .github/workflows/ci.yml:
```diff
@@ -216,21 +217,21 @@ jobs:
       fail-fast: false
       matrix:
-        java: ['11']
+        java: ['11', '17', '21']
```
Isn't 17 already tested above?

guizmaii replied:
Changed. Thanks :)

guizmaii committed:
Remove Java 17 from `testJvms` CI step. It's already tested in other jobs

guizmaii committed:
Fix compilation

guizmaii commented:
Ready for review

guizmaii committed:
Fix `build-website` CI step

ghostdogpr commented on .github/workflows/ci.yml:
```diff
@@ -282,19 +283,19 @@ jobs:
+        uses: actions/setup-java@v4.2.1
         with:
           distribution: temurin
-          java-version: 8
```
That change triggers a compile error when trying to compile code depending on zio with Java 11 (I got it with the latest snapshot):
```
[error] Uncaught exception when running tests: java.lang.UnsupportedClassVersionError: zio/internal/MutableQueueFieldsPadding has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
```
I think we have to compile with the lowest supported Java version. Can you fix it?

ghostdogpr replied:
Opened a PR https://github.com/zio/zio/pull/8780

guizmaii replied:
My bad! Thanks 🙂
